### PR TITLE
Set transaction isolation to read_uncommitted due to @Async

### LINKED ISSUE #1798: 评论后发送邮件报错
### 是什么版本出现了此问题？

1.5.1

### 使用的什么数据库？

MySQL 5.7

### 使用的哪种方式部署？

Fat Jar

### 在线站点地址

_No response_

### 发生了什么？

多次留言发送邮件会报错
现象：
设置好邮件接收
增加1条评论 ---> 收到邮件提醒
再增加1条评论 ---> 无邮件提醒
再增加1条评论 ---> 无邮件提醒
**前往宝塔后台将反代缓存打开后再关闭**
再增加1条评论 ---> 收到邮件提醒
再增加1条评论 ---> 无邮件提醒
再增加1条评论 ---> 无邮件提醒
**前往宝塔后台将反代缓存打开**
再增加1条评论 ---> 无邮件提醒
再增加1条评论 ---> 无邮件提醒
**前往宝塔后台将反代缓存关闭**
再增加1条评论 ---> 无邮件提醒
再增加1条评论 ---> 无邮件提醒

测试环境：
Windows Server 2019，Nginx 1.12.1，Halo 1.5.0 - 出现
CentOS，Nginx 1.12.1，Halo 1.5.0 - 出现
CentOS，Nginx 1.12.1，Halo 1.5.1 - 出现

一般将反代缓存打开后再关闭就能收一次邮件，Windows每次都可以，但是CentOS的偶尔会失效

### 相关日志输出

```shell
[2m2022-03-31 15:13:09.524[0;39m [33m WARN[0;39m [35m3108[0;39m [2m---[0;39m [2m[        Timer-0][0;39m [36mrun.halo.app.cache.AbstractCacheStore   [0;39m [2m:[0;39m Cache key: [cache_lock_public run.halo.app.model.dto.BaseCommentDTO run.halo.app.controller.content.api.SheetController.comment(run.halo.app.model.params.SheetCommentParam):223.74.22.15] has been expired
[2m2022-03-31 15:24:55.683[0;39m [31mERROR[0;39m [35m3108[0;39m [2m---[0;39m [2m[        task-67][0;39m [36m.a.i.SimpleAsyncUncaughtExceptionHandler[0;39m [2m:[0;39m Unexpected exception occurred invoking async method: public void run.halo.app.listener.comment.CommentEventListener.handleCommentNewEvent(run.halo.app.event.comment.CommentNewEvent)

run.halo.app.exception.NotFoundException: SheetComment was not found or has been deleted
        at run.halo.app.service.base.AbstractCrudService.lambda$getById$0(AbstractCrudService.java:139) ~[classes!/:1.5.0]
        at java.base/java.util.Optional.orElseThrow(Optional.java:408) ~[na:na]
        at run.halo.app.service.base.AbstractCrudService.getById(AbstractCrudService.java:138) ~[classes!/:1.5.0]
        at run.halo.app.service.base.AbstractCrudService$$FastClassBySpringCGLIB$$d181db6d.invoke(<generated>) ~[classes!/:1.5.0]
        at org.springframework.cglib.proxy.MethodProxy.invoke(MethodProxy.java:218) ~[spring-core-5.3.16.jar!/:5.3.16]
        at org.springframework.aop.framework.CglibAopProxy$DynamicAdvisedInterceptor.intercept(CglibAopProxy.java:689) ~[spring-aop-5.3.16.jar!/:5.3.16]
        at run.halo.app.service.impl.SheetCommentServiceImpl$$EnhancerBySpringCGLIB$$4a383dee.getById(<generated>) ~[classes!/:1.5.0]
        at run.halo.app.listener.comment.CommentEventListener.handleCommentNewEvent(CommentEventListener.java:135) ~[classes!/:1.5.0]
        at run.halo.app.listener.comment.CommentEventListener$$FastClassBySpringCGLIB$$7fe2191c.invoke(<generated>) ~[classes!/:1.5.0]
        at org.springframework.cglib.proxy.MethodProxy.invoke(MethodProxy.java:218) ~[spring-core-5.3.16.jar!/:5.3.16]
        at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.invokeJoinpoint(CglibAopProxy.java:783) ~[spring-aop-5.3.16.jar!/:5.3.16]
        at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:163) ~[spring-aop-5.3.16.jar!/:5.3.16]
        at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.proceed(CglibAopProxy.java:753) ~[spring-aop-5.3.16.jar!/:5.3.16]
        at org.springframework.aop.interceptor.AsyncExecutionInterceptor.lambda$invoke$0(AsyncExecutionInterceptor.java:115) ~[spring-aop-5.3.16.jar!/:5.3.16]
        at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264) ~[na:na]
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) ~[na:na]
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) ~[na:na]
        at java.base/java.lang.Thread.run(Thread.java:834) ~[na:na]

[2m2022-03-31 15:25:09.520[0;39m [33m WARN[0;39m [35m3108[0;39m [2m---[0;39m [2m[        Timer-0][0;39m [36mrun.halo.app.cache.AbstractCacheStore   [0;39m [2m:[0;39m Cache key: [cache_lock_public run.halo.app.model.dto.BaseCommentDTO run.halo.app.controller.content.api.SheetController.comment(run.halo.app.model.params.SheetCommentParam):223.74.22.15] has been expired
[2m2022-03-31 15:25:17.537[0;39m [32m INFO[0;39m [35m3108[0;39m [2m---[0;39m [2m[pool-4-thread-5][0;39m [36mrun.halo.app.mail.AbstractMailService   [0;39m [2m:[0;39m Sent an email to [[接收邮箱]] successfully, subject: [您的博客页面《关于页面》有了 新的评论。], sent date: [Thu Mar 31 15:25:17 CST 2022]
[2m2022-03-31 15:25:28.996[0;39m [33m WARN[0;39m [35m3108[0;39m [2m---[0;39m [2m[p1296928719-267][0;39m [36mrun.halo.app.cache.AbstractCacheStore   [0;39m [2m:[0;39m Cache key: [cache_lock_public run.halo.app.model.dto.BaseCommentDTO run.halo.app.controller.content.api.SheetController.comment(run.halo.app.model.params.SheetCommentParam):223.74.22.15] has been expired
[2m2022-03-31 15:25:28.996[0;39m [31mERROR[0;39m [35m3108[0;39m [2m---[0;39m [2m[        task-69][0;39m [36m.a.i.SimpleAsyncUncaughtExceptionHandler[0;39m [2m:[0;39m Unexpected exception occurred invoking async method: public void run.halo.app.listener.comment.CommentEventListener.handleCommentNewEvent(run.halo.app.event.comment.CommentNewEvent)

run.halo.app.exception.NotFoundException: SheetComment was not found or has been deleted
        at run.halo.app.service.base.AbstractCrudService.lambda$getById$0(AbstractCrudService.java:139) ~[classes!/:1.5.0]
        at java.base/java.util.Optional.orElseThrow(Optional.java:408) ~[na:na]
        at run.halo.app.service.base.AbstractCrudService.getById(AbstractCrudService.java:138) ~[classes!/:1.5.0]
        at run.halo.app.service.base.AbstractCrudService$$FastClassBySpringCGLIB$$d181db6d.invoke(<generated>) ~[classes!/:1.5.0]
        at org.springframework.cglib.proxy.MethodProxy.invoke(MethodProxy.java:218) ~[spring-core-5.3.16.jar!/:5.3.16]
        at org.springframework.aop.framework.CglibAopProxy$DynamicAdvisedInterceptor.intercept(CglibAopProxy.java:689) ~[spring-aop-5.3.16.jar!/:5.3.16]
        at run.halo.app.service.impl.SheetCommentServiceImpl$$EnhancerBySpringCGLIB$$4a383dee.getById(<generated>) ~[classes!/:1.5.0]
        at run.halo.app.listener.comment.CommentEventListener.handleCommentNewEvent(CommentEventListener.java:135) ~[classes!/:1.5.0]
        at run.halo.app.listener.comment.CommentEventListener$$FastClassBySpringCGLIB$$7fe2191c.invoke(<generated>) ~[classes!/:1.5.0]
        at org.springframework.cglib.proxy.MethodProxy.invoke(MethodProxy.java:218) ~[spring-core-5.3.16.jar!/:5.3.16]
        at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.invokeJoinpoint(CglibAopProxy.java:783) ~[spring-aop-5.3.16.jar!/:5.3.16]
        at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:163) ~[spring-aop-5.3.16.jar!/:5.3.16]
        at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.proceed(CglibAopProxy.java:753) ~[spring-aop-5.3.16.jar!/:5.3.16]
        at org.springframework.aop.interceptor.AsyncExecutionInterceptor.lambda$invoke$0(AsyncExecutionInterceptor.java:115) ~[spring-aop-5.3.16.jar!/:5.3.16]
        at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264) ~[na:na]
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) ~[na:na]
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) ~[na:na]
        at java.base/java.lang.Thread.run(Thread.java:834) ~[na:na]
```

### 附加信息

_No response_

### PR DESCRIPTION
**Pull request by JohnNiang**
### What this PR does

Set transaction isolation to read_uncommitted because the comment event handlers are executed asynchronously.

### Why we need it?

Comment events handlers are executed asynchronously.
And default transaction isolation is read_committed,
so the handlers can not see the comments created by event dispatcher

### Which issue this PR fixes

Fix #1798 

### References

- http://www.h2database.com/html/advanced.html#transaction_isolation
- https://dev.mysql.com/doc/refman/8.0/en/innodb-transaction-isolation-levels.html
- https://docs.spring.io/spring-framework/docs/current/javadoc-api/org/springframework/transaction/event/TransactionalEventListener.html

/kind bug
/area core
/milestone 1.5.x

### CODE REVIEW

**Review by ruibaby — COMMENTED** [review]
/lgtm

File: src/main/java/run/halo/app/listener/comment/CommentEventListener.java [hunk 1]
```diff
@@ -94,6 +96,7 @@ public CommentEventListener(MailService mailService, OptionService optionService
      */
     @Async
     @EventListener
+    @Transactional(isolation = Isolation.READ_UNCOMMITTED)
```
Comment (guqing) [inline src/main/java/run/halo/app/listener/comment/CommentEventListener.java:99]: @JohnNiang  是否可以使用`@TransactionalEventListener`来代替读未提交呢

Comment (JohnNiang) [reply src/main/java/run/halo/app/listener/comment/CommentEventListener.java]: 又学到了不少 :+1:

**Comment by JohnNiang** [comment]
/hold

I'll test if that method is executed asynchronously.

**Comment by JohnNiang** [comment]
That works very well. Please see the screenshot below:

image

/unhold

**Review by guqing — COMMENTED** [review]
/lgtm

**Review by ruibaby — COMMENTED** [review]
/approve

**Pull request merged**